# Recitation 02：从表达式到完整程序

**依据 Lecture 02 与 Recitation 02；中文讲解，保留英文原题。**

完成本讲义后，你应能：

- 先判断表达式的计算顺序，再同时说出结果和 Python 类型。
- 分清字符串字面量、变量和类型转换，解释正常输出或报错的原因。
- 跟踪多重赋值，区分 `print` 的内容、分隔符和结束符。
- 拆读格式说明符，并把百分比、温度换算和配方比例写成输入 → 处理 → 输出程序。

建议 **75–90 分钟**：概念复习 15 分钟；题 1–2 15 分钟；题 3–4 15 分钟；题 5–6 15 分钟；题 7–9 20 分钟；回顾 5 分钟。自学时按小节暂停，先预测、再运行、最后用一句话解释“为什么”。课堂时间不足时，题 6 的组合格式和应用题的额外测试可留作课后练习。

教师版包含逐项答案、易错点和已验证输出；学生版保留同一组原题、概念提示和可运行的 TODO。学生版初始 TODO 单元格故意不完成计算，可直接 **Restart Kernel and Run All**；填题后再运行，才能检验自己的答案。应用题使用可编辑的输入字符串，避免运行 notebook 时等待键盘；对应行旁边给出 `input()` 替换方式。

**来源版本与页码：** Lecture 02 是原生 PPTX，共 55 张 slide；引用中的 Slide N 是演示文稿顺序。Recitation 02 是 4 页 PDF。书籍为 Gaddis *Starting Out With Python*, Revel 6th ed.；`PDF p.` 一律指从 1 开始的物理页，`printed p.` 才是书内印刷页。逐题引用与哈希保存在同目录的 [source-map.md](source-map.md) / [source-map.json](source-map.json)。`direct` 表示直接讲到；`prerequisite` 表示只提供必要背景，具体组合由本讲义推导。

**原资料提示：** Lecture 02 Slide 55 写的是 “Read chapter 3”，但本次练习主要对应书籍 Chapter 2；这里保留阅读要求并标出不一致，不替老师更改作业。Lecture Slides 20、24、33、49 有个别演算或注释笔误；本讲义的演示以可运行 Python 为准，具体核对见教师说明。题 2 的后续章节引用只用于查 `str()` 和重复语法，不要求学习文件、列表或循环。

## 热身 A：先看运算符，再看类型

整数通常是 `int`，带小数点的数通常是 `float`。`/` 做真正的除法，即使两个整数能整除，结果也为 `float`。`//` 向负无穷取整；这与 `int()` 把小数部分朝 0 去掉不同。对这里的非零除数，余数满足 `a == (a // b) * b + (a % b)`；结果不为 0 时，余数与除数同号。

先算括号，再算幂，再算乘除、整除、取余，最后加减。同一优先级的乘除取余从左往右。不要只写数字而忘了类型：一项出现浮点数可能改变最终类型。下面是热身例子，数值与原题不同。

In [1]:
print(29 / 6, type(29 / 6).__name__)
print(29 // 6, 29 % 6)
print(-29 // 6, int(-29 / 6))  # 向下取整 vs 朝 0 截断
print(4 + 6 * 2, (4 + 6) * 2)


4.833333333333333 float
4 5
-5 -4
16 20


## 热身 B：值、名字与类型

`"8"` 是字符串，`8` 是整数。字符串的 `+` 连接字符串；数值的 `+` 做加法。Python 不会替你猜测应该把哪一边转换。`int()`、`float()` 把合适的字符串变成数；`str()` 把值变成字符串。字符串重复需要整数次数。

变量名不用引号；有引号时得到的是那几个字符本身。`x = y` 把当时的 `y` 值赋给 `x`，不会建立以后自动同步的公式。多重赋值先求出整个右边的旧值，再按左边名字的顺序绑定。

In [2]:
text_count = "8"
print(text_count + "2", int(text_count) + 2)
print("la" * 3)
x = 9
y = x
x = 2
print(x, y)
x, y = y, x  # 先得到旧的右边，再赋给左边
print(x, y)


82 10
lalala
2 9
9 2


## 热身 C：字符串边界与输出边界

字符串开头的引号必须由同类引号结束；另一种引号可以作为内容。`\'` 和 `\"` 表示引号字符，`\n` 是换行，`\t` 是制表符。制表符移动到下一个制表位，显示距离受当前位置和环境影响，**不是固定几个空格**。

`print(value1, value2, sep=..., end=...)` 先把值变成文本，再在相邻参数之间放 `sep`，最后加 `end`。默认 `sep=' '`、`end='\n'`；字符串里面原有的空格不会被删除。`repr()` 是观察工具：它用带引号、带转义的形式描述文本，这些外层引号不是原始 `print` 自动打印的内容。

In [3]:
print("red", "blue", sep=" / ", end="!\n")
message = "X\tY\nZ"
print(message)
print("辅助 repr：", repr(message))


red / blue!
X	Y
Z
辅助 repr： 'X\tY\nZ'


## 热身 D：格式化是生成显示文本

f-string 的 `{表达式:格式说明}` 先计算值，再生成字符串。字段宽度是**最小**宽度，内容太长通常不会被剪掉。`<` 左对齐，`>` 右对齐；紧挨对齐符之前的字符可以作为填充。

精度取决于被格式化的类型：浮点数的 `.2f` 表示小数点后两位；浮点数只写 `.2`、省略 `f` 时，使用类似通用格式的有效数字规则，不等于固定两位小数。字符串的 `.2` 最多保留两个字符。先读值的类型，再解释精度；不要把宽度当精度。格式化不改变原变量的数值。

In [4]:
amount = 6.789
print("|" + f"{amount:9.2f}" + "|")
print("|" + f"{'owl':.>7}" + "|")
print("|" + f"{'sparrow':<7.2}" + "|")
print(amount)  # 原来的值没有变


|     6.79|
|....owl|
|sp     |
6.789


## 热身 E：输入 → 处理 → 输出

`input()` 总是返回字符串。先按意义选 `int()`（整数数量）或 `float()`（允许小数），再计算，最后打印结果和单位。先用能心算的输入检查公式方向和比例。

In [5]:
distance_text = "3.5"  # 交互时替换为 input("距离（km）：")
distance_km = float(distance_text)
distance_m = distance_km * 1000
print(f"距离：{distance_m:.1f} m")


距离：3500.0 m


**教师辅助工具（无需学生记忆）。** 下方工具仅把实际打印捕获为 `repr`，帮助检查换行、制表符、前后空格与 `end`。练习答案中的原始语句仍会正常执行。这里的 `import`、函数、上下文管理器与 `exec` 是验证设施，不是本次要求掌握的语法；学生版会移除工具及其调用。报错演示使用 `try/except` 或 `compile`，让 notebook 继续执行；捕获后显示的 `error — ...` 是讲义的提示，不是原题语句产生的正常输出。

In [6]:
import io
from contextlib import redirect_stdout

def show_invisible(source, values=None):
    """仅供教师：执行同一段原题，显示实际标准输出的 repr。"""
    capture = io.StringIO()
    with redirect_stdout(capture):
        exec(source, {} if values is None else values.copy())
    print("\n[辅助 repr；外层引号不是原输出]", repr(capture.getvalue()))


## 1. 数值表达式：结果、类型和运算顺序

**原题（保留来源提取文本）**

```text
Exercises 1 – Numeric values and operations 
Evaluate the following expressions and specify their type.  
5 * -6 
 
34 / 10 
 
(34 // 10) * (1.0 + 2) 
 
34 % 10 
 
23 % 4.0 
 
25 + 30 / 6 
 
-7 % 3 
 
100 - 25 * 3 % 4
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 14 (PPTX), “Mixing numeric types” — **direct**；覆盖：numeric result types。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 19 (PPTX), “Operators on int and float” — **direct**；覆盖：true and floor division / remainder operation。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 20 (PPTX), “Operators on int and float” — **direct**；覆盖：negative-operand remainder。 This slide includes a negative-dividend modulo example; its separate 10 // 0.25 output contains a typo.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 22 (PPTX), “Operators Precedence” — **direct**；覆盖：operator precedence。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 90 (PDF p. 91) — **direct**；覆盖：numeric result types。 Read the general mixed-type rules together with the explicit / exception on PDF p.84.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 83 (PDF p. 84) — **direct**；覆盖：true and floor division。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 87 (PDF p. 88) — **direct**；覆盖：remainder operation。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 83 (PDF p. 84) — **prerequisite**；覆盖：negative-operand remainder。 Teaches negative floor division, a prerequisite for interpreting negative modulo; does not explicitly demonstrate negative-dividend %.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 84 (PDF p. 85) — **direct**；覆盖：operator precedence。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=1)，PDF p. 1（物理页）。

### 本题考察什么

这题同时检查“怎么算”和“算出来是什么类型”。8 个表达式依次涉及：① 负数参与乘法；② 真除法的返回类型；③ 括号、整除与混合类型；④ 整数余数；⑤ 浮点除数下的余数类型；⑥ 除法优先于加法；⑦ 负数的整除—余数关系；⑧ 乘法、取余、减法的先后。

**怎么做：** 每项先在纸上写出结果和类型，再运行；解释至少一步中间计算。

**来源与表述说明：** 负数取余的教材引用是 prerequisite：教材讲负数整除，本讲义再用商和余数的关系推导。教材 PDF p. 91 的一般类型规则要与 PDF p. 84 的 `/` 例外一起读。

### 1.1 `5 * -6`

预测：结果 = ___；类型 = ___。

In [7]:
value = 5 * -6
print(value, type(value).__name__)


-30 int


**逐项讲解：** 两个操作数都是整数；先理解 `-6` 是负数，再相乘，得到 `-30`，类型为 `int`。易错：不要把负号误当成乘法之后才执行的另一条语句。

### 1.2 `34 / 10`

预测：结果 = ___；类型 = ___。

In [8]:
value = 34 / 10
print(value, type(value).__name__)


3.4 float


**逐项讲解：** `/` 做真正的除法，结果 `3.4`，类型为 `float`。即便能整除，`/` 的结果仍是 `float`，不是“有小数才变 float”。

### 1.3 `(34 // 10) * (1.0 + 2)`

预测：结果 = ___；类型 = ___。

In [9]:
value = (34 // 10) * (1.0 + 2)
print(value, type(value).__name__)


9.0 float


**逐项讲解：** 左括号 `34 // 10` 得到整数 `3`；右括号 `1.0 + 2` 得到浮点数 `3.0`；最终 `9.0`，类型为 `float`。易错：只看到整除就把整个表达式判为 int。

### 1.4 `34 % 10`

预测：结果 = ___；类型 = ___。

In [10]:
value = 34 % 10
print(value, type(value).__name__)


4 int


**逐项讲解：** `34 = 3 * 10 + 4`，余数 `4`，类型为 `int`。`%` 在这里不是“百分比符号”。

### 1.5 `23 % 4.0`

预测：结果 = ___；类型 = ___。

In [11]:
value = 23 % 4.0
print(value, type(value).__name__)


3.0 float


**逐项讲解：** `23 = 5 * 4.0 + 3.0`，结果 `3.0`，类型为 `float`。余数即使没有非零小数部分，也不会因此自动变回 int。

### 1.6 `25 + 30 / 6`

预测：结果 = ___；类型 = ___。

In [12]:
value = 25 + 30 / 6
print(value, type(value).__name__)


30.0 float


**逐项讲解：** 先算 `30 / 6 = 5.0`，再算 `25 + 5.0 = 30.0`，类型为 `float`。不能把 `25 + 30` 先加起来。

### 1.7 `-7 % 3`

预测：结果 = ___；类型 = ___。

In [13]:
value = -7 % 3
print(value, type(value).__name__)


2 int


**逐项讲解：** 先求向下取整的商：`-7 // 3 = -3`，不是朝 0 截断的 `-2`。再由 `-7 = (-3) * 3 + 2` 得到余数 `2`，类型为 `int`。这不是把正数余数直接加上负号。

### 1.8 `100 - 25 * 3 % 4`

预测：结果 = ___；类型 = ___。

In [14]:
value = 100 - 25 * 3 % 4
print(value, type(value).__name__)


97 int


**逐项讲解：** 乘法与取余同级，先从左到右算 `25 * 3 = 75`，再算 `75 % 4 = 3`，最后 `100 - 3 = 97`，类型为 `int`。易错：先做减法，或者把 `%` 当成更低优先级。

## 2. 看输出：同一个符号，不同的类型

**原题（保留来源提取文本）**

```text
Exercise 2 - Looking for Prints  
Write out the output of each line in the space provided. If the line results in an error, 
write "error". 
print(5 + "5")  
 
 
    print(5 + int("5")) 
print(str(5) + int("5"))    
    print(15 % 2) 
print("hey" + "yo")  
 
    print(2 * "go") 
print("yo" * 4.0) 
 
 
    print("hi" + there) 
print("go" * 2 + "!" * 2)
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 15 (PPTX), “Data Type Conversion” — **direct**；覆盖：int conversion / str conversion。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 19 (PPTX), “Operators on int and float” — **direct**；覆盖：numeric remainder。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 25 (PPTX), “Operations Summary” — **direct**；覆盖：string concatenation and operand types。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 24 (PPTX), “Operator on String” — **direct**；覆盖：string repetition / integer repetition count。 The slide directly shows both operand orders and a float-count TypeError; its separate Hello + ICP example has an incorrect displayed result.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 33 (PPTX), “Variables” — **direct**；覆盖：unassigned variable names。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：int conversion。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 6, §6.1: Introduction to File Input and Output, printed p. 362 (PDF p. 363) — **direct**；覆盖：str conversion。 Targeted later-chapter reference for str() only; file handling is outside this recitation.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 87 (PDF p. 88) — **direct**；覆盖：numeric remainder。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.8: String Concatenation, printed p. 93 (PDF p. 94) — **direct**；覆盖：string concatenation and operand types。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 8, §8.3: Testing, Searching, and Manipulating Strings, printed p. 500 (PDF p. 501) — **direct**；覆盖：string repetition。 Targeted later-chapter reference for string repetition only; loops and password validation are outside this recitation.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 7, §7.2: Introduction to Lists, printed p. 406 (PDF p. 407) — **prerequisite**；覆盖：integer repetition count。 States the integer operand requirement for sequence repetition using lists; string repetition is separately demonstrated on PDF p.501.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 68 (PDF p. 69) — **direct**；覆盖：unassigned variable names。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=1)，PDF p. 1（物理页）。

### 本题考察什么

这题检查你能否在运行前判断操作数类型。按原表格逐行、从左到右共有 9 项：① 数值和字符串能否相加；② `int` 转换后的加法；③ `str` 和 `int` 混用；④ 数值取余；⑤ 字符串连接；⑥ 整数重复次数；⑦ 看起来像整数的浮点次数；⑧ 带引号文本与未赋值名字；⑨ 重复与连接的组合顺序。

**怎么做：** 写正常输出或 error，并说明是哪一类类型/名字问题。把 9 行视为独立表达式；`there` 未预先定义。

**来源与表述说明：** 右列行首空格来自 PDF 表格排版，不是 Python 缩进。后续教材章节只定位个别语法；列表重复的引用是 integer repetition count 的 prerequisite。

### 2.1

```python
print(5 + "5")
```

预测输出或 error，并写出判断依据。

In [15]:
try:
    print(5 + "5")
except TypeError as error:
    print("error — TypeError:", error)


error — TypeError: unsupported operand type(s) for +: 'int' and 'str'


**逐项讲解：** `5` 是 int，`"5"` 是 str；两者不能直接用 `+`。结果为 error（TypeError），没有正常打印值。

### 2.2

```python
print(5 + int("5"))
```

预测输出或 error，并写出判断依据。

In [16]:
print(5 + int("5"))


10


**逐项讲解：** `int("5")` 得到整数 5，数值相加后输出 `10`。转换的是值，不是把 `int` 这个词打印出来。

### 2.3

```python
print(str(5) + int("5"))
```

预测输出或 error，并写出判断依据。

In [17]:
try:
    print(str(5) + int("5"))
except TypeError as error:
    print("error — TypeError:", error)


error — TypeError: can only concatenate str (not "int") to str


**逐项讲解：** 左边经过 `str` 后为字符串，右边经过 `int` 后为整数，仍是 str + int，得到 error（TypeError）。“两边都转换过”并不表示类型一致。

### 2.4

```python
print(15 % 2)
```

预测输出或 error，并写出判断依据。

In [18]:
print(15 % 2)


1


**逐项讲解：** 数值取余，15 除以 2 余 1，因此输出 `1`。

### 2.5

```python
print("hey" + "yo")
```

预测输出或 error，并写出判断依据。

In [19]:
print("hey" + "yo")


heyyo


**逐项讲解：** 两个字符串直接连接，输出 `heyyo`，不会自动加空格；字符串分隔引号不属于输出。

### 2.6

```python
print(2 * "go")
```

预测输出或 error，并写出判断依据。

In [20]:
print(2 * "go")


gogo


**逐项讲解：** 整数在左侧也可以指定字符串重复次数，输出 `gogo`。

### 2.7

```python
print("yo" * 4.0)
```

预测输出或 error，并写出判断依据。

In [21]:
try:
    print("yo" * 4.0)
except TypeError as error:
    print("error — TypeError:", error)


error — TypeError: can't multiply sequence by non-int of type 'float'


**逐项讲解：** 虽然 4.0 的数学值是整数，它的 Python 类型仍是 float；字符串重复次数要求整数，得到 error（TypeError）。

### 2.8

```python
print("hi" + there)
```

预测输出或 error，并写出判断依据。

In [22]:
try:
    print("hi" + there)
except NameError as error:
    print("error — NameError:", error)


error — NameError: name 'there' is not defined


**逐项讲解：** 先求变量 `there` 的值，但它未定义，所以得到 error（NameError）。不能擅自把 `there` 当成字符串字面量。

### 2.9

```python
print("go" * 2 + "!" * 2)
```

预测输出或 error，并写出判断依据。

In [23]:
print("go" * 2 + "!" * 2)


gogo!!


**逐项讲解：** 先分别重复得到 `gogo` 和 `!!`，再连接，输出 `gogo!!`。没有额外空格。

## 3. 引号与转义：代码文本和实际字符

**原题（保留来源提取文本）**

```text
Exercises 3 - Escape sequence 
Write the output, or Error if an error occurs. 
print("Welcome to \'ICP\'") 
 
print("Welcome to "ICP") 
 
print('Welcome to "ICP') 
print('Welcome to \"ICP') 
 
print("Welcome to \"ICP\"") 
 
print("Today is\t\"Tuesday\"\nWelcome to\t\"ICP\"")
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 9 (PPTX), “Strings” — **direct**；覆盖：string delimiters and syntax errors。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 10 (PPTX), “Strings” — **direct**；覆盖：escaped quotation marks。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 11 (PPTX), “Escape Sequence” — **direct**；覆盖：newline and tab characters。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.3: Displaying Output with the print Function, printed p. 63 (PDF p. 64) — **prerequisite**；覆盖：string delimiters and syntax errors。 Teaches matching delimiters and mixed quote styles; does not explicitly trace the malformed print line in this exercise.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 97 (PDF p. 98) — **direct**；覆盖：escaped quotation marks。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 96 (PDF p. 97) — **direct**；覆盖：newline and tab characters。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=1)，PDF p. 1（物理页）。

### 本题考察什么

6 行分别检查：① 双引号字符串内转义单引号；② 字符串定界符是否匹配；③ 单引号字符串内的普通双引号；④ 同一位置的转义双引号；⑤ 成对显示双引号；⑥ 引号、制表符、换行共同出现时的实际输出。判断错误时先检查语法，再讨论输出。

**怎么做：** 先数清字符串从哪里开始、在哪里结束；再把转义序列解释成实际字符。第 3、4 行虽然同处原表格一格，也分别完成。

### 3.1

```python
print("Welcome to \'ICP\'")
```

In [24]:
print("Welcome to \'ICP\'")
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print("Welcome to \\\'ICP\\\'")')


Welcome to 'ICP'

[辅助 repr；外层引号不是原输出] "Welcome to 'ICP'\n"


**逐项讲解：** 双引号界定整个字符串，`\'` 产生一个单引号。实际输出 `Welcome to 'ICP'`；反斜杠不显示。此处单引号不需要转义也能作为内容，但原题写法有效。

### 3.2

```python
print("Welcome to "ICP")
```

In [25]:
bad_code = 'print("Welcome to "ICP")'
try:
    compile(bad_code, "<exercise-3-line-2>", "exec")
except SyntaxError as error:
    print("error — SyntaxError:", error.msg)


error — SyntaxError: unterminated string literal (detected at line 1)


**逐项讲解：** 中间的双引号过早结束字符串，后面的 `ICP` 和末尾引号无法组成合法表达式，发生 SyntaxError。它不是运行后才产生的类型问题，原题没有正常输出。讲义用 compile 捕获错误，不改写原题。

### 3.3

```python
print('Welcome to "ICP')
```

In [26]:
print('Welcome to "ICP')
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print(\'Welcome to "ICP\')')


Welcome to "ICP

[辅助 repr；外层引号不是原输出] 'Welcome to "ICP\n'


**逐项讲解：** 外层用单引号，里面的双引号是普通字符，因此输出 `Welcome to "ICP`，只有左边那一个双引号；不要自行补上原题没写的右双引号。

### 3.4

```python
print('Welcome to \"ICP')
```

In [27]:
print('Welcome to \"ICP')
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print(\'Welcome to \\"ICP\')')


Welcome to "ICP

[辅助 repr；外层引号不是原输出] 'Welcome to "ICP\n'


**逐项讲解：** `\"` 同样得到双引号字符，因此这一行与上一行输出完全相同。反斜杠改变源代码写法，不会作为字符留在输出中。

### 3.5

```python
print("Welcome to \"ICP\"")
```

In [28]:
print("Welcome to \"ICP\"")
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print("Welcome to \\"ICP\\"")')


Welcome to "ICP"

[辅助 repr；外层引号不是原输出] 'Welcome to "ICP"\n'


**逐项讲解：** 两处 `\"` 各产生一个双引号，所以输出 `Welcome to "ICP"`。代码最外层的字符串引号不显示，内容里的两个双引号会显示。

### 3.6

```python
print("Today is\t\"Tuesday\"\nWelcome to\t\"ICP\"")
```

In [29]:
print("Today is\t\"Tuesday\"\nWelcome to\t\"ICP\"")
# 教师观察辅助：显示同一 print 的实际输出，含末尾换行。
show_invisible('print("Today is\\t\\"Tuesday\\"\\nWelcome to\\t\\"ICP\\"")')


Today is	"Tuesday"
Welcome to	"ICP"

[辅助 repr；外层引号不是原输出] 'Today is\t"Tuesday"\nWelcome to\t"ICP"\n'


**逐项讲解：** 第一行文本含 Tuesday，换行后第二行含 ICP；每行相应位置各有一个制表符。`\n` 是内容中的换行，print 的默认 end 还会在整个输出末尾补一个换行。repr 中的 `\t` 不能解释成固定数量的空格。

## 4. 追踪 Lady：多重赋值先读旧值

**原题（保留来源提取文本）**

```text
Exercise 4 | Variable assignment 
Consider the following variable assignments:  
 
Left = "Lady" 
 
Middle = "Jack" 
 
Right = "Ace"  
Figure out where the 'Lady' is (Left, Middle, Right) after running the following programs: 
Left, Right, Middle = Left, Middle, Right 
 
Left, Right, Middle = Middle, Right, Left 
 
Left, Middle, Right = Middle, Right, Left 
 
Left, Middle, Right = Middle, Right, Left 
Left, Middle, Right = Middle, Right, Left 
 
Left, Middle, Right = Middle, Right, Left 
Left, Middle, Right = Middle, Right, Left 
Left, Middle, Right = Middle, Right, Left 
 
Left, Middle, Right = Right, Right, Right 
Left, Middle, Right = Left, Left, Left 
Left, Middle, Right = Middle, Middle, Middle
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 33 (PPTX), “Variables” — **direct**；覆盖：multiple assignment。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 29 (PPTX), “How do variables actually work?” — **prerequisite**；覆盖：old right-hand-side values before rebinding。 Teaches evaluate-then-bind for ordinary assignment; applying that order to every RHS value in a multiple assignment requires an explicit teaching step.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 32 (PPTX), “Reassignment of Variables” — **direct**；覆盖：reassignment and loss of references。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 69 (PDF p. 70) — **direct**；覆盖：multiple assignment。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 69 (PDF p. 70) — **prerequisite**；覆盖：old right-hand-side values before rebinding。 Demonstrates multiple assignment of literals; does not explicitly explain taking all old RHS values before rebinding existing names.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 71 (PDF p. 72) — **direct**；覆盖：reassignment and loss of references。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=2)，PDF p. 2（物理页）。

### 本题考察什么

这题考察多重赋值、左右位置的配对、逐行状态追踪，以及重新赋值后原值是否仍由某个变量保存。六段程序分别有 1、1、1、2、3、3 条赋值语句；不要把表格所有行连成一段程序。

**怎么做：** 每段都从 `Left="Lady"`、`Middle="Jack"`、`Right="Ace"` 开始。对每条语句先写右侧旧值，再按左侧出现顺序分配；最后回答 Lady 在哪里。如果没有变量再保存 Lady，可以回答“无”，无需勉强选三个位置之一。

**来源与表述说明：** 原题只列 Left/Middle/Right 选项；这里补充“无”作为必要的判定范围，不提前指定适用于哪段。关于先求全部 RHS 再绑定名字的引用包含 prerequisite，本讲义明确补充这一步推理。

### 4.1 独立程序（1 条赋值）

```python
Left, Right, Middle = Left, Middle, Right
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

In [30]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Right, Middle = Left, Middle, Right
print("第 1 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Lady Ace Jack


**逐项讲解：** 先取右侧旧值 `(Lady, Jack, Ace)`，但左侧顺序是 `(Left, Right, Middle)`，最终按固定列 `(Left, Middle, Right)` 看为 `(Lady, Ace, Jack)`；Lady 在 Left。易错：忽略左侧名字的顺序。

### 4.2 独立程序（1 条赋值）

```python
Left, Right, Middle = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

In [31]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Right, Middle = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Lady Ace


**逐项讲解：** 先取右侧旧值 `(Jack, Ace, Lady)`，对应左侧 `(Left, Right, Middle)`，最终固定列为 `(Jack, Lady, Ace)`；Lady 在 Middle。易错：把第三个值误放进 Right。

### 4.3 独立程序（1 条赋值）

```python
Left, Middle, Right = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

In [32]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Ace Lady


**逐项讲解：** 右侧旧值为 `(Jack, Ace, Lady)`，左侧这次是标准顺序，最终 `(Jack, Ace, Lady)`；Lady 在 Right。不要用已更新的 Left 去计算右侧最后一项。

### 4.4 独立程序（2 条赋值）

```python
Left, Middle, Right = Middle, Right, Left
Left, Middle, Right = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

In [33]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 2 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Ace Lady
第 2 行后: Ace Lady Jack


**逐项讲解：** 第 1 行后 `(Jack, Ace, Lady)`，第 2 行以这个新状态作为旧值，得到 `(Ace, Lady, Jack)`；Lady 在 Middle。每行内部同时绑定，行与行之间按顺序执行。

### 4.5 独立程序（3 条赋值）

```python
Left, Middle, Right = Middle, Right, Left
Left, Middle, Right = Middle, Right, Left
Left, Middle, Right = Middle, Right, Left
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

In [34]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 1 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 2 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Right, Left
print("第 3 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Jack Ace Lady
第 2 行后: Ace Lady Jack
第 3 行后: Lady Jack Ace


**逐项讲解：** 三行后依次为 `(Jack, Ace, Lady)` → `(Ace, Lady, Jack)` → `(Lady, Jack, Ace)`；Lady 回到 Left。这是同一个循环移位做三次，而不是三次都读取最初的值。

### 4.6 独立程序（3 条赋值）

```python
Left, Middle, Right = Right, Right, Right
Left, Middle, Right = Left, Left, Left
Left, Middle, Right = Middle, Middle, Middle
```

追踪列始终按 `Left | Middle | Right` 记录，最终 Lady 的位置：___。

In [35]:
Left = "Lady"
Middle = "Jack"
Right = "Ace"
print("初始 (Left, Middle, Right):", Left, Middle, Right)
Left, Middle, Right = Right, Right, Right
print("第 1 行后:", Left, Middle, Right)
Left, Middle, Right = Left, Left, Left
print("第 2 行后:", Left, Middle, Right)
Left, Middle, Right = Middle, Middle, Middle
print("第 3 行后:", Left, Middle, Right)


初始 (Left, Middle, Right): Lady Jack Ace
第 1 行后: Ace Ace Ace
第 2 行后: Ace Ace Ace
第 3 行后: Ace Ace Ace


**逐项讲解：** 第 1 行把当时 Right 的 Ace 同时赋给三个变量，因此状态成为 `(Ace, Ace, Ace)`；后两行仍然只能复制 Ace。Lady 已不在任何变量中，答案为“无”。变量数没有减少，但它们保存的值可能相同，原先的值不会自动保留下来。

## 5. print 的 sep 与 end：逐字符拼接

**原题（保留来源提取文本）**

```text
Exercise 5 | Print arguments  
a = 5 
print("a", a, sep=' = ') 
 
a = 5 
print("a = ", a, sep="0000") 
 
a = 5 
print("a = ", a, sep="'", end="'") 
 
a, b = 'b', 'a' 
print(a, 'b', sep='>>>', end='\t') 
print('a', b, sep='***', end='\n\n') 
print(a, b, sep='\t')
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：string literals versus variable values。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 46 (PPTX), “print – sep and end arguments” — **direct**；覆盖：print separators and endings。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 11 (PPTX), “Escape Sequence” — **direct**；覆盖：newline and tab characters。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.5: Variables, printed p. 71 (PDF p. 72) — **direct**；覆盖：string literals versus variable values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 95 (PDF p. 96) — **direct**；覆盖：print separators and endings。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.9: More About the print Function, printed p. 96 (PDF p. 97) — **direct**；覆盖：newline and tab characters。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=3)，PDF p. 3（物理页）。

### 本题考察什么

这题把字符串/变量、分隔符和结束符放在一起。4 段程序依次检查：① 字面量与变量之间自定义分隔；② 原字符串尾空格和分隔字符串叠加；③ 分隔符与结束符中的引号是真实字符；④ 多次 print 因不同 end 连成哪些行。

**怎么做：** 原题以表格留白呈现，讲义将任务解读为“预测每段程序的完整输出”。每段独立执行，包含最后一段的全部 3 次 print；明确最后是否有换行。

**来源与表述说明：** 本题没有额外英文指令句；上面的中文任务说明是教学解释，不是增补伪造的原题文字。

### 5.1

```python
a = 5
print("a", a, sep=' = ')
```

In [36]:
a = 5
print("a", a, sep=' = ')
show_invisible('a = 5\nprint("a", a, sep=\' = \')')


a = 5

[辅助 repr；外层引号不是原输出] 'a = 5\n'


**逐项讲解：** 第一个参数是字符 a，第二个参数是变量值 5；只在两者之间插入 ` = `，输出 `a = 5`，然后默认换行。

### 5.2

```python
a = 5
print("a = ", a, sep="0000")
```

In [37]:
a = 5
print("a = ", a, sep="0000")
show_invisible('a = 5\nprint("a = ", a, sep="0000")')


a = 00005

[辅助 repr；外层引号不是原输出] 'a = 00005\n'


**逐项讲解：** 第一个参数 `"a = "` 自带末尾空格，再插入四个 0，之后才是 5，完整文本为 `a = 00005`，然后换行。sep 不会替换或删除参数内的空格。

### 5.3

```python
a = 5
print("a = ", a, sep="'", end="'")
```

In [38]:
a = 5
print("a = ", a, sep="'", end="'")
show_invisible('a = 5\nprint("a = ", a, sep="\'", end="\'")')


a = '5'
[辅助 repr；外层引号不是原输出] "a = '5'"


**逐项讲解：** 参数 `"a = "` 后有空格，sep 插入单引号，再输出 5，end 再加单引号，所以完整输出为 `a = '5'`，末尾没有换行。这两个单引号是输出内容；辅助 repr 外层的引号则只是描述记号。

### 5.4

```python
a, b = 'b', 'a'
print(a, 'b', sep='>>>', end='\t')
print('a', b, sep='***', end='\n\n')
print(a, b, sep='\t')
```

In [39]:
a, b = 'b', 'a'
print(a, 'b', sep='>>>', end='\t')
print('a', b, sep='***', end='\n\n')
print(a, b, sep='\t')
show_invisible("a, b = 'b', 'a'\nprint(a, 'b', sep='>>>', end='\\t')\nprint('a', b, sep='***', end='\\n\\n')\nprint(a, b, sep='\\t')")


b>>>b	a***a

b	a

[辅助 repr；外层引号不是原输出] 'b>>>b\ta***a\n\nb\ta\n'


**逐项讲解：** 变量 a 保存 b，变量 b 保存 a。第 1 次输出 `b>>>b`，以 tab 结束；第 2 次紧接着输出 `a***a`，随后两个换行造成一空行；第 3 次输出 b、tab、a，再默认换行。完整 repr 是 `'b>>>b\ta***a\n\nb\ta\n'`。不要把变量名字当成其当前值，也不要假设每次 print 都单独占一行。

## 6. 格式说明符：宽度、填充、对齐与精度

**原题（保留来源提取文本）**

```text
Exercise 6 | String formatting 
print(f"{12:8d}") 
 
print(f"{'12':0>8}") 
 
num = 12.34567890 
print(f"{num:8.3f}") 
 
print(f"{12.34567890:<08.3f}") 
 
name = 'dolphin' 
res1 = f'{name:->10}' 
res2 = f'{name:<10.3}' 
print(res1) 
print(res2)
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 48 (PPTX), “f-string” — **direct**；覆盖：f-string placeholders。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 53 (PPTX), “f-string - number formatting” — **direct**；覆盖：numeric field width and precision。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 51 (PPTX), “f-string – string formatting” — **direct**；覆盖：alignment / custom fill characters。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 52 (PPTX), “f-string – string formatting” — **direct**；覆盖：string precision truncation。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 53 (PPTX), “f-string - number formatting” — **prerequisite**；覆盖：zero padding with explicit alignment。 Shows zero-padded numeric fields without an explicit alignment character. Slide 51 supplies alignment background; the combination <08.3f is not directly demonstrated.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 98 (PDF p. 99) — **direct**；覆盖：f-string placeholders。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 100 (PDF p. 101) — **direct**；覆盖：numeric field width and precision。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 102 (PDF p. 103) — **direct**；覆盖：numeric field width and precision。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 105 (PDF p. 106) — **direct**；覆盖：alignment。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 105 (PDF p. 106) — **prerequisite**；覆盖：custom fill characters / zero padding with explicit alignment。 Teaches alignment and width, but not arbitrary fill characters, zero padding, or zero padding combined with explicit alignment.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.10: Displaying Formatted Output with F-strings, printed p. 99 (PDF p. 100) — **prerequisite**；覆盖：string precision truncation。 Teaches the format-specifier grammar; this textbook section does not explain string precision as truncation.

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=3)，PDF p. 3（物理页）。

### 本题考察什么

5 段程序分别检查：① 整数格式 d 与最小字段宽度；② 字符串配自定义填充及右对齐；③ 浮点数固定小数位和宽度；④ 显式左对齐与零填充的组合；⑤ 字符串右对齐填充，以及左对齐、截断和末尾空格。最后一段有两个输出，两个都要预测。

**怎么做：** 原题只有表格，讲义将任务解读为预测输出。先写值的类型，再依次标注填充、对齐、宽度、精度、格式类型；用竖线边界或 repr 检查不可见空格。

**来源与表述说明：** 自定义填充、字符串精度、零填充结合显式对齐的教材引用含 prerequisite；它们提供语法背景而不直接给出原题组合。教材 PDF p. 103 将 D 列为整数格式的说法有误，本题只使用有效的 d。

### 6.1

```python
print(f"{12:8d}")
```

In [40]:
print(f"{12:8d}")
show_invisible('print(f"{12:8d}")')


      12

[辅助 repr；外层引号不是原输出] '      12\n'


**逐项讲解：** `12` 是 int，`d` 用十进制整数显示；`8` 是最小总宽度，不是补 8 个字符。数字本身长 2，默认右对齐，所以前面补 6 个空格：`'      12'`。print 再追加换行。

### 6.2

```python
print(f"{'12':0>8}")
```

In [41]:
print(f"{'12':0>8}")
show_invisible('print(f"{\'12\':0>8}")')


00000012

[辅助 repr；外层引号不是原输出] '00000012\n'


**逐项讲解：** `'12'` 是 str；`0` 是填充字符，`>` 是右对齐，`8` 是最小总宽度。左侧补 6 个 0，得到 `'00000012'`。这里的 0 是显示字符，不是改变字符串的数值类型。

### 6.3

```python
num = 12.34567890
print(f"{num:8.3f}")
```

In [42]:
num = 12.34567890
print(f"{num:8.3f}")
show_invisible('num = 12.34567890\nprint(f"{num:8.3f}")')


  12.346

[辅助 repr；外层引号不是原输出] '  12.346\n'


**逐项讲解：** 先按 `.3f` 舍入到小数点后 3 位，得到 `12.346`（6 个字符，含小数点）；再按宽度 8 右对齐，前面补 2 个空格：`'  12.346'`。字段宽度包含整数、小数点与小数部分。

### 6.4

```python
print(f"{12.34567890:<08.3f}")
```

In [43]:
print(f"{12.34567890:<08.3f}")
show_invisible('print(f"{12.34567890:<08.3f}")')


12.34600

[辅助 repr；外层引号不是原输出] '12.34600\n'


**逐项讲解：** `<` 明确左对齐，`0` 启用零填充，`8` 是宽度，`.3f` 仍然只负责舍入到 3 位小数。先得到 `12.346`，再在右边补 2 个 0 到宽度 8，结果为 `'12.34600'`。末尾的两个 0 是填充，不是多计算了两位小数；不能把结果解释为精度变成了 5。

### 6.5

```python
name = 'dolphin'
res1 = f'{name:->10}'
res2 = f'{name:<10.3}'
print(res1)
print(res2)
```

In [44]:
name = 'dolphin'
res1 = f'{name:->10}'
res2 = f'{name:<10.3}'
print(res1)
print(res2)
show_invisible("name = 'dolphin'\nres1 = f'{name:->10}'\nres2 = f'{name:<10.3}'\nprint(res1)\nprint(res2)")


---dolphin
dol       

[辅助 repr；外层引号不是原输出] '---dolphin\ndol       \n'


**逐项讲解：** 第一个格式 `->10`：用 `-` 填充、右对齐、宽度 10，dolphin 长 7，所以输出 `'---dolphin'`。第二个格式 `<10.3` 针对 str：`.3` 最多保留前三个字符 `dol`，再左对齐到宽度 10，右边补 7 个空格，输出 `'dol       '`。字符串 `.3` 是截断字符；浮点 `.3f` 是三位小数，浮点省略 f 的 `.3` 又是另一种有效数字规则。

## 7. 销售利润预测：百分比是乘数

**原题（保留来源提取文本）**

```text
Exercise 7 | Sales Prediction  
A company has determined that its annual profit is typically 23 percent of total sales. 
Write a program that asks the user to enter the projected amount of total sales, and 
then displays the profit that will be made from that amount.
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 40 (PPTX), “Input function” — **direct**；覆盖：numeric keyboard input。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 49 (PPTX), “f-string” — **direct**；覆盖：percentage calculation。 The example relates a fractional multiplier to a percentage; its printed f-string quotes and :2f explanatory label have typos.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：displaying computed values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：numeric keyboard input。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 82 (PDF p. 83) — **direct**；覆盖：percentage calculation / displaying computed values。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=4)，PDF p. 4（物理页）。

### 本题考察什么

这题要求把一句商业描述转成输入 → 处理 → 输出：读取可含小数的总销售额，把字符串转换成数值，用给定利润比例计算，并清楚显示金额。关键是百分数的含义，和“销售额”与“利润”两个量的区别。

**怎么做：** 使用可编辑的 `sales_text` 模拟 input 的字符串返回值，补全转换、计算和输出。完成后把该行改成注释所示 input，即为交互程序；暂不要求处理不合法输入。

In [45]:
sales_text = "10000"  # 交互版：sales_text = input("预计总销售额：")
total_sales = float(sales_text)
profit = total_sales * 0.23
print(f"预计利润：${profit:.2f}")


预计利润：$2300.00


**推导与易错点：** 23% 表示每 100 单位销售额对应 23 单位利润，所以乘数是 `23 / 100 = 0.23`。输入先经过 `float` 才能进行数值计算。总销售额 10000 对应利润 2300，显示为 `$2300.00`。不要乘 23，也不要把总销售额乘 1.23；后者是把原数增加 23%，与题意不同。`.2f` 只控制金额显示，不应先把销售额转成 int 丢掉小数。

## 8. 摄氏转华氏：翻译数学公式

**原题（保留来源提取文本）**

```text
Exercise 8 | Celsius to Fahrenheit Temperature Converter 
Write a program that converts Celsius temperatures to Fahrenheit temperatures. The 
formula is as follows: 
𝐹= 9
5 𝐶+ 32 
The program should ask the user to enter a temperature in Celsius, then display the 
temperature converted to Fahrenheit.
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 40 (PPTX), “Input function” — **direct**；覆盖：numeric keyboard input。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 19 (PPTX), “Operators on int and float” — **prerequisite**；覆盖：translating a mathematical formula。 Teaches the arithmetic operators needed to implement the supplied formula, but does not teach Celsius conversion or the translation of mathematical notation.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：displaying computed values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：numeric keyboard input。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 88 (PDF p. 89) — **direct**；覆盖：translating a mathematical formula。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 82 (PDF p. 83) — **direct**；覆盖：displaying computed values。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=4)，PDF p. 4（物理页）。

### 本题考察什么

这题检查输入转换、数学分式的程序表达、乘法符号必须显式写出，以及输出单位。你需要保留公式中的比例关系与偏移量，不能改变运算的组合。

**怎么做：** 先读下面恢复排版的公式，再用可编辑输入完成换算，并用容易心算的温度检验。

**来源与表述说明：** 原 PDF 将 9/5 排成上下分式，文本提取成为两行；上面的原题保留原始提取文本。下方仅恢复已核对的公式排版。

**恢复原分式的排版：**

$$F = \frac{9}{5}C + 32$$

其中 $C$ 是摄氏温度，$F$ 是华氏温度。这不是 `9 / (5*C + 32)`；写代码前先确认分数线的范围。

In [46]:
celsius_text = "25"  # 交互版：celsius_text = input("摄氏温度：")
celsius = float(celsius_text)
fahrenheit = (9 / 5) * celsius + 32
print(f"{celsius:.1f} °C = {fahrenheit:.1f} °F")


25.0 °C = 77.0 °F


**推导与易错点：** 先把 `9/5` 作为比例，乘摄氏温度，再加 32；在 Python 中相邻的 `(9 / 5)` 和 `celsius` 之间必须写 `*`。`/` 会得到浮点比值 1.8，不能用 `//` 代替。示例 25 °C 对应 77 °F。可检查 0 °C → 32 °F、100 °C → 212 °F、−40 °C → −40 °F；这三个值同时帮助发现漏加 32、比例写反、误用整除等错误。

## 9. 饼干配方调整：同一个比例用于三种原料

**原题（保留来源提取文本）**

```text
Exercise 9 | Ingredient Adjuster   
A cookie recipe calls for the following ingredients:  
1) 1.5 cups of sugar 
2) 1 cup of butter 
3) 2.75 cups of flour  
The recipe produces 48 cookies with this amount of the ingredients. 
Write a program that asks the user how many cookies he or she wants to make, and 
then displays the number of cups of each ingredient needed for the specified number of 
cookies.
```

**Knowledge references / 知识点出处**

- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 40 (PPTX), “Input function” — **direct**；覆盖：numeric keyboard input。
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 36 (PPTX), “You Try it!” — **prerequisite**；覆盖：proportional ingredient scaling。 Uses price-times-quantity as proportional calculation background; it does not derive scaling from a 48-cookie reference recipe.
- Lecture: [*Lecture 02*](../../lecture/Lecture%2002.pptx), Slide 44 (PPTX), “print function” — **direct**；覆盖：displaying computed values。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.6: Reading Input from the Keyboard, printed p. 78 (PDF p. 79) — **direct**；覆盖：numeric keyboard input。
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 88 (PDF p. 89) — **prerequisite**；覆盖：proportional ingredient scaling。 Teaches multiplication, division, and formula translation, but does not derive the recipe scaling factor. The matching original textbook exercise is on PDF p.147, printed p.146.
- Textbook: [Gaddis, 6th ed.](../../textbook/Starting%20Out%20With%20Python%20--%20Revel%206th%20Edition%20%28Tony%20Gaddis%29.pdf), Chapter 2, §2.7: Performing Calculations, printed p. 82 (PDF p. 83) — **direct**；覆盖：displaying computed values。

原题来源：[recitation/Recitaion 02.pdf](../../recitation/Recitaion%2002.pdf#page=4)，PDF p. 4（物理页）。

### 本题考察什么

这题检查比例建模和变量复用。先从目标饼干数量与基准产量的关系得到缩放比例，再分别计算糖、黄油和面粉用量，并统一显示 cups 单位。题目里的 1)、2)、3) 是同一个配方的三种原料，不是三个独立应用。

**怎么做：** 先运行一次共同输入单元格，再按糖 → 黄油 → 面粉顺序完成三个计算单元格。改变目标数量后，从共同输入开始重跑本题；最终要得到一份含三种配料的完整结果。

**来源与表述说明：** 第 3 个编号在源块中带有配方产量与总体要求，这些共同要求已完整保留在原题中；三个编号共享同一目标数量与缩放比例。

**程序组织：** 先读目标数量，再确定缩放比例，最后依次输出糖、黄油、面粉。48 块是基准产量；目标数量应为非负整数，输入合法性检查留给后续课程。下面四个代码单元格合起来是一份顺序执行的程序。

In [47]:
cookies_text = "24"  # 交互版：cookies_text = input("想制作多少块饼干：")
cookies = int(cookies_text)
scale = cookies / 48
print(f"制作 {cookies} 块饼干需要：")


制作 24 块饼干需要：


**共同推导：** 基准配方产出 48 块，所以目标配方是原配方的 `cookies / 48` 倍。目标 24 时比例为 0.5，目标 96 时比例为 2.0。比例必须使用 `/`，`//` 会把少于一批的正比例截成 0。输入数量用 int 合理，但配料量必须允许小数。

### 9.1 糖（原题第 1 项）

使用共同缩放比例计算糖，并标明 cups。

In [48]:
sugar = 1.5 * scale
print(f"糖：{sugar:.3f} cups")


糖：0.750 cups


**本小项推导：** 糖的基准用量为 1.5 cups，因此乘共同的 scale。24 块时需要 0.750 cups；不要按增加/减少多少块去直接加减 1.5。

### 9.2 黄油（原题第 2 项）

用同一目标数量与比例计算黄油。

In [49]:
butter = 1 * scale
print(f"黄油：{butter:.3f} cups")


黄油：0.500 cups


**本小项推导：** 黄油基准是 1 cup，因此用量为 `1 * scale`。24 块时是 0.500 cups。不要因为原始用量是整数就用 int 转换结果，那会丢失半杯。

### 9.3 面粉（原题第 3 项）

完成最后一种原料，再检查三行输出是否对应同一批饼干。

In [50]:
flour = 2.75 * scale
print(f"面粉：{flour:.3f} cups")


面粉：1.375 cups


**本小项推导：** 面粉的基准用量为 2.75 cups，同样乘 scale。24 块需要 1.375 cups。三种原料必须同时缩放；`.3f` 是为了清楚显示这个例子的杯数，不是题目规定的强制格式。改变输入后要从本题共同输入单元格重跑，不能让某种配料沿用旧比例。

**教师：复制到 `.py` 后可直接交互的一份完整程序。** 此文本代码块不会在 Run All 时等待输入。

```python
cookies = int(input("想制作多少块饼干："))
scale = cookies / 48
sugar = 1.5 * scale
butter = 1 * scale
flour = 2.75 * scale
print(f"制作 {cookies} 块饼干需要：")
print(f"糖：{sugar:.3f} cups")
print(f"黄油：{butter:.3f} cups")
print(f"面粉：{flour:.3f} cups")
```

**代表值核对：** 24 块（半批）是 0.750 / 0.500 / 1.375 cups；48 块是 1.500 / 1.000 / 2.750 cups；96 块（双批）是 3.000 / 2.000 / 5.500 cups，顺序均为糖 / 黄油 / 面粉。0 块应全部为 0。不要只核对总和；每一种配料都应满足比例。

## 回顾与离堂检查

现在合上答案，用自己的话解释：

1. 为什么表达式的结果数字和 Python 类型都要检查？`/` 和 `//` 的差别是什么？
2. 遇到输出题时，你会按什么顺序检查类型、变量名、引号、`sep` 和 `end`？
3. 多重赋值的一行里，何时读取右侧旧值？两行之间又如何更新？
4. 字段宽度、数值精度与字符串精度分别控制什么？如何检查末尾空格？
5. 一个应用程序中，输入值、计算得到的值和显示格式有什么区别？你会选什么容易心算的输入检验公式？

**自查方法：** 能预测只是第一步；还要能说明每项为什么得到这个结果。若结果不符，先找第一步发生分歧的位置。完成 TODO 后重启内核、从头运行，确认程序不依赖上一次留下的变量。

**教师资料核对：原文件保持不变。**

- Lecture Slide 20：`10 // 0.25` 应为 `40.0`（浮点参与整除仍为 float）。
- Slide 24：`"Hello" + "ICP"` 应为 `HelloICP`，连接不自动插入字符。
- Slide 33：`z = 1` 后的跟随注释有误；Python 名字区分大小写，相关注释不可照抄。
- Slide 49：f-string 示例引号不匹配；固定两位小数需要 `.2f`，`:2f` 的 2 是宽度位置，不是两位小数精度。
- Textbook PDF p. 91 的概括要结合 p. 84 `/` 的类型例外；PDF p. 103 的大写 `D` 不是有效整数格式。本次只教实际使用的 `d`。

追问建议：让学生给出判断规则，再运行验证。E4 不要把“看见三个变量”误当成“保存着三个不同值”。E6 先解释各位置含义，再讨论组合规则；后续章节引用只供定位，不扩展为列表、文件或循环课。